python/ray/train/ml_trainer.py

```diff
@@ -0,0 +1,108 @@
+import abc
+import logging
+from typing import Dict, Union, Callable, Optional
+
+from ray.data import Dataset
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.ml.result import Result
+from ray import tune
+from ray.tune.trainable import ConvertibleToTrainable
+from ray.tune.utils import deep_update
+
+GenDataset = Union[Dataset, Callable[[], Dataset]]
+
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: support conditional hyperparameter tuning of Trainer __init__ args.
+class Trainer(ConvertibleToTrainable, abc.ABC):
+    """
+    Defines interface for distributed training on Ray.
+
+    Args:
+        scaling_config (Optional[ScalingConfig]): Configuration for how to
+            scale training.
+        run_config (Optional[RunConfig]): Configuration for the execution of
+            the training run.
+        train_dataset (Optional[GenDataset]): Either a distributed Ray
+            :ref:`Dataset <dataset-api>` or a Callbable that returns a Dataset,
+            to use for training. If a ``preprocessor`` is also provided,
+            it will be fit on this dataset and this dataset will be
+            transformed.
+        additional_datasets (Optional[GenDataset]): Any additional
```

Comment:
```suggestion
        extra_datasets (Optional[GenDataset]): Any extra
```

Shorter / easier to spell?

Reply:
Updated!